graylog2-server/src/main/java/org/graylog2/shared/buffers/processors/ProcessBufferProcessor.java

```diff
@@ -171,6 +173,7 @@ private void handleMessage(@Nonnull Message msg) {
 
             message.getStreams().forEach(s -> streamMetrics.markIncomingMeter(s.getId()));
             message.ensureValidTimestamp();
+            message.normalizeTimestamp(configuration.getTimestampGracePeriod());
```

Comment:
When the value is coming from the cluster config, we need to make sure we do not load it on every message.
Instead we need to ensure that we listen to changes and cache the value, as this is a very hot code path.

Reply:
`ensureValidTimestamp` contains similar code, can we unify the handling of the checks somehow to not add unnecessary overhead in a hot code path?

Reply:
I kept this separate, because `ensureValidTimestamp` is called frequently via `getTimeStamp`. We only need to normalize once, not on every read operation.